qos/evm/service_state.go

```diff
@@ -0,0 +1,278 @@
+		logger.Warn().Err(err).Msg("error filtering endpoints")
+		return protocol.EndpointAddr(""), err
+	}
+
+	if len(filteredEndpointsAddr) == 0 {
+		logger.Error().Msg("all endpoints failed validation; selecting a random endpoint.")
+		randomAvailableEndpointAddr := availableEndpoints[rand.Intn(len(availableEndpoints))]
+		return randomAvailableEndpointAddr, nil
+	}
+
+	logger.With(
+		"total_endpoints", len(availableEndpoints),
+		"endpoints_after_filtering", len(filteredEndpointsAddr),
+	).Info().Msg("filtered endpoints")
+
+	// TODO_FUTURE: consider ranking filtered endpoints, e.g. based on latency, rather than randomization.
+	selectedEndpointAddr := filteredEndpointsAddr[rand.Intn(len(filteredEndpointsAddr))]
+	return selectedEndpointAddr, nil
+}
+
+// filterValidEndpoints returns the subset of available endpoints that are valid
+// according to previously processed observations.
+func (ss *serviceState) filterValidEndpoints(availableEndpoints []protocol.EndpointAddr) ([]protocol.EndpointAddr, error) {
+	ss.endpointStore.endpointsMu.RLock()
+	defer ss.endpointStore.endpointsMu.RUnlock()
+
+	logger := ss.logger.With("method", "filterValidEndpoints").With("qos_instance", "evm")
+
+	if len(availableEndpoints) == 0 {
+		return nil, errors.New("received empty list of endpoints to select from")
+	}
+
+	logger.Info().Msg(fmt.Sprintf("About to filter through %d available endpoints", len(availableEndpoints)))
+
+	// TODO_FUTURE: use service-specific metrics to add an endpoint ranking method
+	// which can be used to assign a rank/score to a valid endpoint to guide endpoint selection.
+	var filteredEndpointsAddr []protocol.EndpointAddr
+	for _, availableEndpointAddr := range availableEndpoints {
+		logger := logger.With("endpoint_addr", availableEndpointAddr)
+		logger.Info().Msg("processing endpoint")
+
+		endpoint, found := ss.endpointStore.endpoints[availableEndpointAddr]
+		if !found {
+			logger.Info().Msg(fmt.Sprintf("endpoint %s not found in the store. Skipping...", availableEndpointAddr))
+			continue
+		}
+
+		if err := ss.validateEndpoint(endpoint); err != nil {
+			logger.Info().Err(err).Msg(fmt.Sprintf("skipping endpoint that failed validation: %v", endpoint))
+			continue
+		}
+
+		filteredEndpointsAddr = append(filteredEndpointsAddr, availableEndpointAddr)
+		logger.Info().Msg(fmt.Sprintf("endpoint %s passed validation", availableEndpointAddr))
+	}
+
+	return filteredEndpointsAddr, nil
+}
+
+// validateEndpoint returns an error if the supplied endpoint is not
+// valid based on the perceived state of the EVM blockchain.
+//
+// It returns an error if:
+// - The endpoint has returned an empty response in the past.
+// - The endpoint's response to an `eth_chainId` request is not the expected chain ID.
+// - The endpoint's response to an `eth_blockNumber` request is greater than the perceived block number.
+// - The endpoint's archival check is invalid, if enabled.
+func (ss *serviceState) validateEndpoint(endpoint endpoint) error {
```

Comment:
Great refactor: consolidating endpoint validation logic in serviceState. This will help make `endpoint` merely a data container.